Docs: clarify how to fix "Too many files (>8) have matched the default project"

### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

I have been unable to upgrade typescript-eslint past 7.5.0; doing so gives a "Too many files have matched the default project" whenever I try to lint.


My languageOptions are

```
languageOptions: {
    parser: tseslint.parser,
    parserOptions: {
      EXPERIMENTAL_useProjectService: true,
    },
  },
```

The error I'm getting:

```
common:lint: <redacted path>
common:lint:   0:0  error  Parsing error: Too many files (>8) have matched the default project.
common:lint: 
common:lint: Having many files run with the default project is known to cause performance issues and slow down linting.
common:lint: 
common:lint: See https://typescript-eslint.io/troubleshooting/typed-linting#allowdefaultprojectforfiles-glob-too-wide
common:lint: 
common:lint: Matching files:
common:lint: - /<redacted path>/node_modules/tinycolor2/esm/tinycolor.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/endOfMonth/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/lastDayOfMonth/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/startOfDay/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/startOfISOWeek/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/startOfMonth/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/subDays/index.js
common:lint: - /<redacted path>/node_modules/date-fns/esm/subMonths/index.js
common:lint: - /<redacted path>/node_modules/sinon/pkg/sinon-esm.js
common:lint: 
common:lint: If you absolutely need more files included, set parserOptions.EXPERIMENTAL_useProjectService.maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING to a larger value
```

I've read various issues and docs and none of them answer my question.

- The [FAQ](https://typescript-eslint.io/troubleshooting/typed-linting/#i-get-errors-telling-me-having-many-files-run-with-the-default-project-is-known-to-cause-performance-issues-and-slow-down-linting) tells me I've configured `allowDefaultProjectForFiles` with too wide a glob. This isn't true; I haven't configured the property at all and I can't find what its default is anywhere.
- I found https://github.com/typescript-eslint/typescript-eslint/issues/9032 but it was closed and the issue still persists when I try to use the latest 7.x.



I'm certainly open to the possibility that something is wrong with my linting configuration, but I couldn't find anything in the docs to help me investigate this. (Are `node_modules` files supposed to be opened by typescript-eslint?)

From my perspective, it looks like a breaking change was released and it would be nice to at least clarify in the docs how folks with a previously working v7 config can upgrade.

### Affected URL(s)

https://typescript-eslint.io/troubleshooting/typed-linting/#i-get-errors-telling-me-having-many-files-run-with-the-default-project-is-known-to-cause-performance-issues-and-slow-down-linting

Hey, I've wanted to add a bit more to this issue. I have the same error, however I believe my ESLint setup is correct (using the popular Antfu ESLint config with a few tweaks).

I've wanted to use tsconfig project references by enabling `parserOptions.EXPERIMENTAL_useProjectService` and I've got the same error as @JavaScriptBach, but pointing out actual source-code files and not `node_modules` files:

```
/<redacted>/src/components/ui/VueSelect.vue
  0:0  error  Parsing error: Too many files (>8) have matched the default project.

Having many files run with the default project is known to cause performance issues and slow down linting.

See https://typescript-eslint.io/troubleshooting/typed-linting#allowdefaultprojectforfiles-glob-too-wide

Matching files:
- /<redacted>/src/App.vue
- /<redacted>/cypress/support/e2e.ts
- /<redacted>/cypress/support/commands.ts
- /<redacted>/src/components/faq-accordion/FaqAccordion.vue
- /<redacted>/src/views/HomeView.vue
- /<redacted>/src/components/faq-accordion/FaqAccordionSlideItem.vue
- /<redacted>/src/components/header/HeaderAwards.vue
- /<redacted>/src/components/SiteFooter.vue
- /<redacted>/src/components/header/HeaderStars.vue
- /<redacted>/src/components/header/HeaderBackground.vue
- /<redacted>/src/components/icons/AcademicCapIcon.vue
- /<redacted>/src/components/faq-accordion/FaqAccordionButton.vue
- /<redacted>/src/components/faq-accordion/FaqAccordionSlide.vue
- /<redacted>/src/components/icons/StarHalfFilledIcon.vue
- /<redacted>/src/components/icons/StarIcon.vue
- /<redacted>/src/components/ui/Logo.vue
- /<redacted>/src/components/ui/VueSelect.vue

If you absolutely need more files included, set parserOptions.EXPERIMENTAL_useProjectService.maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING to a larger value
```

I agree that using the following code works, but seems hacky and the official documentation is lacking about possible properties for the `EXPERIMENTAL_useProjectService` object:

```js
parserOptions: {
  EXPERIMENTAL_useProjectService: {
    maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING: 1000
  },
}
```

Is there a better workaround than setting `maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING` at the moment? Does it will affect _a lot_ the linting speed/performance?

Note: I cannot use a `tsconfig.eslint.json` for type-aware linting in my case, as it was causing an issue. I've noticed that using `EXPERIMENTAL_useProjectService` fixes the issue, that's why I'm using it.

--- a/docs/packages/Parser.mdx
+++ b/docs/packages/Parser.mdx
@@ -326,7 +326,7 @@ This option brings two main benefits over the older `project`:
 For more information, see:
 
 - [feat(typescript-estree): add EXPERIMENTAL_useProjectService option to use TypeScript project service](https://github.com/typescript-eslint/typescript-eslint/pull/6754)
-- [docs: blog post on parserOptions.projectService](docs: blog post on parserOptions.projectService )
+- [docs: blog post on parserOptions.projectService](https://github.com/typescript-eslint/typescript-eslint/pull/8031)
 
 #### `ProjectServiceOptions`
 

--- a/docs/troubleshooting/typed-linting/index.mdx
+++ b/docs/troubleshooting/typed-linting/index.mdx
@@ -25,32 +25,26 @@ See [ESLint does not re-compute cross-file information on file changes (microsof
 
 ## I get errors telling me "Having many files run with the default project is known to cause performance issues and slow down linting."
 
-These errors are caused by using the [`EXPERIMENTAL_useProjectService`](../../packages/Parser.mdx#experimental_useprojectservice) `allowDefaultProject` with an excessively wide glob.
-`allowDefaultProject` causes a new TypeScript "program" to be built for each "out of project" file it includes, which incurs a performance overhead for each file.
+These errors are caused by attempting to use the [`projectService`](../../packages/Parser.mdx#projectservice) to lint a file not explicitly included in a `tsconfig.json`.
 
-To resolve this error, narrow the glob(s) used for `allowDefaultProject` to include fewer files.
-For example:
+For each file being reported:
 
-```diff title="eslint.config.js"
-parserOptions: {
-  EXPERIMENTAL_useProjectService: {
-    allowDefaultProject: [
--      "**/*.js",
-+      "./*.js"
-    ]
-  }
-}
-```
+- If you **do not** want to lint the file:
+  - Use [one of the options ESLint offers to ignore files](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code), such an `ignores` config key.
+- If you **do** want to lint the file:
+  - If you **do not** want to lint the file with [type-aware linting](../../getting-started/Typed_Linting.mdx): [disable type-checked linting for that file](#how-do-i-disable-type-checked-linting-for-a-file).
+  - If you **do** want to lint the file with [type-aware linting](../../getting-started/Typed_Linting.mdx):
+    1. If possible, add the file to the closest `tsconfig.json`'s `include`. For example, allowing `.js` files:
+       ```diff title="tsconfig.json"
+       "include": [
+         "src",
+       + "*.js"
+       ]
+       ```
+    2. Otherwise, consider setting [`parserOptions.createProjectService.allowDefaultProject`](../../packages/parser#allowdefaultproject).
 
-You may also need to include more files in your `tsconfig.json`.
-For example:
-
-```diff title="tsconfig.json"
-"include": [
-  "src",
-+ "*.js"
-]
-```
+typescript-eslint allows up to 8 "out of project" files by default.
+Each file causes a new TypeScript "program" to be built for each file it includes, which incurs a performance overhead _for each file_.
 
 If you cannot do this, please [file an issue on typescript-eslint's typescript-estree package](https://github.com/typescript-eslint/typescript-eslint/issues/new?assignees=&labels=enhancement%2Ctriage&projects=&template=07-enhancement-other.yaml&title=Enhancement%3A+%3Ca+short+description+of+my+proposal%3E) telling us your use case and why you need more out-of-project files linted.
 Be sure to include a minimal reproduction we can work with to understand your use case!
@@ -61,54 +55,17 @@ These errors are caused by an ESLint config requesting type information be gener
 
 ### Fixing the Error
 
-<!-- prettier-ignore-start -->
-
 - If you **do not** want to lint the file:
-  - Use [one of the options ESLint offers](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code) to ignore files, namely a `.eslintignore` file, or `ignorePatterns` config.
+  - Use [one of the options ESLint offers to ignore files](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code), namely a `.eslintignore` file, or `ignorePatterns` config.
 - If you **do** want to lint the file:
   - If you **do not** want to lint the file with [type-aware linting](../../getting-started/Typed_Linting.mdx):
-    - Use [ESLint's `overrides` configuration](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns) with our [`disable-type-checked`](../../users/Shared_Configurations.mdx#disable-type-checked) config to disable type checking for just that type of file.
-    <Tabs groupId="eslint-config">
-      <TabItem value="Flat Config">
-
-      ```js title="eslint.config.js"
-      import tseslint from 'typescript-eslint';
-
-      export default tseslint.config(
-        // ... the rest of your config ...
-        {
-          files: ['**/*.js'],
-          extends: [tseslint.configs.disableTypeChecked],
-        },
-      );
-      ```
-
-      </TabItem>
-      <TabItem value="Legacy Config">
-
-      ```js title=".eslintrc.cjs"
-      module.exports = {
-        // ... the rest of your config ...
-        overrides: [
-          {
-            extends: ['plugin:@typescript-eslint/disable-type-checked'],
-            files: ['./**/*.js'],
-          },
-        ],
-      };
-      ```
-
-      </TabItem>
-    </Tabs>
-    - Alternatively to disable type checking for files manually, you can set `parserOptions: { project: false }` to an override for the files you wish to exclude.
+    - Use [ESLint's configuration objects](https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions) with our [`disable-type-checked`](../../users/Shared_Configurations.mdx#disable-type-checked) config to disable type checking for just that type of file.
   - If you **do** want to lint the file with [type-aware linting](../../getting-started/Typed_Linting.mdx):
     - Check the `include` option of each of the TSConfigs that you provide to `parserOptions.project` - you must ensure that all files match an `include` glob, or else our tooling will not be able to find it.
       - If the file is a `.cjs`, `.js`, or `.mjs` file, make sure [`allowJs`](https://www.typescriptlang.org/tsconfig#allowJs) is enabled.
     - If your file shouldn't be a part of one of your existing tsconfigs (for example, it is a script/tool local to the repo), then consider creating a new tsconfig (we advise calling it `tsconfig.eslint.json`) in your project root which lists this file in its `include`. For an example of this, you can check out the configuration we use in this repo:
       - [`tsconfig.eslint.json`](https://github.com/typescript-eslint/typescript-eslint/blob/main/tsconfig.eslint.json)
       - [`eslint.config.mjs`](https://github.com/typescript-eslint/typescript-eslint/blob/main/eslint.config.mjs)
-
-<!-- prettier-ignore-end -->
 
 ### More Details
 
@@ -136,39 +93,46 @@ See our docs on [type aware linting](../../getting-started/Typed_Linting.mdx) fo
 You're using an outdated version of `@typescript-eslint/parser`.
 Update to the latest version to see a more informative version of this error message, explained [above](#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file 'backlink to I get errors telling me ESLint was configured to run ...').
 
-<HiddenHeading id="allowdefaultproject-glob-too-wide" />
+## How do I disable type-checked linting for a file?
 
-## I get errors telling me "Having many files run with the default project is known to cause performance issues and slow down linting."
+Use [ESLint's configuration objects](https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions) with our [`disable-type-checked`](../../users/Shared_Configurations.mdx#disable-type-checked) config to disable type checking for a `files` match that includes that file.
 
-These errors are caused by using the [`EXPERIMENTAL_useProjectService`](../../packages/Parser.mdx#experimental_useprojectservice) `allowDefaultProject` with an excessively wide glob.
-`allowDefaultProject` causes a new TypeScript "program" to be built for each "out of project" file it includes, which incurs a performance overhead for each file.
+For example, to disable type-checked linting on all `.js` files:
 
-To resolve this error, narrow the glob(s) used for `allowDefaultProject` to include fewer files.
-For example:
+<Tabs groupId="eslint-config">
+  <TabItem value="Flat Config">
 
-```diff title="eslint.config.js"
-parserOptions: {
-  EXPERIMENTAL_useProjectService: {
-    allowDefaultProject: [
--      "**/*.js",
-+      "./*.js"
-    ]
-  }
-}
+```js title="eslint.config.js"
+import tseslint from 'typescript-eslint';
+
+export default tseslint.config(
+  // ... the rest of your config ...
+  {
+    files: ['**/*.js'],
+    extends: [tseslint.configs.disableTypeChecked],
+  },
+);
 ```
 
-You may also need to include more files in your `tsconfig.json`.
-For example:
+  </TabItem>
+  <TabItem value="Legacy Config">
 
-```diff title="tsconfig.json"
-"include": [
-  "src",
-+ "*.js"
-]
+```js title=".eslintrc.cjs"
+module.exports = {
+  // ... the rest of your config ...
+  overrides: [
+    {
+      extends: ['plugin:@typescript-eslint/disable-type-checked'],
+      files: ['./**/*.js'],
+    },
+  ],
+};
 ```
 
-If you cannot do this, please [file an issue on typescript-eslint's typescript-estree package](https://github.com/typescript-eslint/typescript-eslint/issues/new?assignees=&labels=enhancement%2Ctriage&projects=&template=07-enhancement-other.yaml&title=Enhancement%3A+%3Ca+short+description+of+my+proposal%3E) telling us your use case and why you need more out-of-project files linted.
-Be sure to include a minimal reproduction we can work with to understand your use case!
+  </TabItem>
+</Tabs>
+
+Alternatively to disable type checking for files manually, you can set [`parserOptions: { project: false }`](../../packages/Parser.mdx#project) to an override for the files you wish to exclude.
 
 ## typescript-eslint thinks my variable is never nullish / is `any` / etc., but that is clearly not the case to me
 

